apollo-backend/src/test/java/io/logz/apollo/BlockerTest.java

```diff
@@ -156,6 +158,45 @@ public void testServiceBlocker() throws Exception {
 
+    @Test
+    public void testEnvironmentsStackBlocker() throws Exception {
+        ApolloTestClient apolloTestClient = Common.signupAndLogin();
+        ApolloTestAdminClient apolloTestAdminClient = Common.getAndLoginApolloTestAdminClient();
+
+        Environment blockedEnvironment = ModelsGenerator.createAndSubmitEnvironment(apolloTestClient);
+        Environment okEnvironment = ModelsGenerator.createAndSubmitEnvironment(apolloTestClient);
+        Service service = ModelsGenerator.createAndSubmitService(apolloTestClient);
+        Stack blockedEnvironmentsStack = ModelsGenerator.createAndSubmitEnvironmentsStack(apolloTestClient, Arrays.asList(blockedEnvironment.getId()));
+        DeployableVersion deployableVersion = ModelsGenerator.createAndSubmitDeployableVersion(apolloTestClient, service);
+
+        createAndSubmitBlocker(apolloTestAdminClient, "unconditional", "{}", null, null, blockedEnvironmentsStack);
+
+        assertThatThrownBy(() -> ModelsGenerator.createAndSubmitDeployment(apolloTestClient, blockedEnvironment, service, deployableVersion)).isInstanceOf(Exception.class)
+                                                                                                                                                    .hasMessageContaining("Deployment is currently blocked");
+
+        ModelsGenerator.createAndSubmitDeployment(apolloTestClient, okEnvironment, service, deployableVersion);
+    }
+
+    @Test
+    public void testServicesStackBlocker() throws Exception {
+        ApolloTestClient apolloTestClient = Common.signupAndLogin();
+        ApolloTestAdminClient apolloTestAdminClient = Common.getAndLoginApolloTestAdminClient();
+
+        Environment environment = ModelsGenerator.createAndSubmitEnvironment(apolloTestClient);
+        Service blockedService = ModelsGenerator.createAndSubmitService(apolloTestClient);
+        Service okService = ModelsGenerator.createAndSubmitService(apolloTestClient);
+        Stack blockedServicesStack = ModelsGenerator.createAndSubmitServicesStack(apolloTestClient, Arrays.asList(blockedService.getId()));
+        DeployableVersion blockedDeployableVersion = ModelsGenerator.createAndSubmitDeployableVersion(apolloTestClient, blockedService);
+        DeployableVersion okDeployableVersion = ModelsGenerator.createAndSubmitDeployableVersion(apolloTestClient, okService);
+
+        createAndSubmitBlocker(apolloTestAdminClient, "unconditional", "{}", null, null, blockedServicesStack);
+
+        assertThatThrownBy(() -> ModelsGenerator.createAndSubmitDeployment(apolloTestClient, environment, blockedService, blockedDeployableVersion)).isInstanceOf(Exception.class)
+                                                                                                                                                    .hasMessageContaining("Deployment is currently blocked");
+
+        ModelsGenerator.createAndSubmitDeployment(apolloTestClient, environment, okService, okDeployableVersion);
+    }
+
```

Comment:
Good tests. Add another that will prevent us from adding an invalid blocker (with stack and environment + with stack and service + with stack and both)